File: sdks/typescript/src/runtime.ts
/**
 * Runtime-side secret substitution helpers.
 *
 * Cloud SDK adapters resolve placeholders at credential-load time so
 * AWS SigV4/GCP JWT/Azure SAS signing sees the real credential before
 * computing request signatures. The vsock transport lands with W3; this
 * registry is the stable SDK contract the transport plugs into.
 */

export type SubstitutionHandler = (placeholder: string) => string | Promise<string>;

const handlers = new Map<string, SubstitutionHandler>();

export class SubstitutionHandlerError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SubstitutionHandlerError";
  }
}

export interface AwsCredentials {
  accessKeyId: string;
  secretAccessKey: string;
  sessionToken?: string;
}

export function register_substitution_handler(
  name: string,
  fn: SubstitutionHandler,
): void {
  if (name.length === 0) {
    throw new Error("substitution handler name must be non-empty");
  }
  handlers.set(name, fn);
}

export function clear_substitution_handlers(): void {
  handlers.clear();
}

export async function substitute(name: string, placeholder: string): Promise<string> {
  const handler = handlers.get(name);
  if (handler === undefined) {
    throw new SubstitutionHandlerError(
      `no substitution handler registered for ${JSON.stringify(name)}`,
    );
  }
  return await handler(placeholder);
}

export async function aws_credentials_from_placeholders(input: {
  accessKeyId: string;
  secretAccessKey: string;
  sessionToken?: string;
}): Promise<AwsCredentials> {
  const creds: AwsCredentials = {
    accessKeyId: await substitute("aws", input.accessKeyId),
    secretAccessKey: await substitute("aws", input.secretAccessKey),
  };
  if (input.sessionToken !== undefined) {
    creds.sessionToken = await substitute("aws", input.sessionToken);
  }
  return creds;
}

export function is_placeholder(value: string): boolean {
  return value.startsWith("mvm-secret://");
}

File: sdks/python/mvm/_runtime.py
"""Runtime-side secret substitution helpers.

This module is intentionally dependency-light: cloud SDK adapters call
``register_substitution_handler`` at credential load time, then run their
normal signing flow with the materialized credential. The actual vsock
transport lands with the W3 substitution service; tests inject handlers
directly so the contract is stable before the transport exists.
"""

from __future__ import annotations

from dataclasses import dataclass
from hmac import compare_digest
from typing import Callable, Dict


SubstitutionHandler = Callable[[str], str]


_handlers: Dict[str, SubstitutionHandler] = {}


class SubstitutionHandlerError(RuntimeError):
    """Raised when credential substitution cannot be performed."""


@dataclass(frozen=True)
class AwsCredentials:
    access_key_id: str
    secret_access_key: str
    session_token: str | None = None


def register_substitution_handler(name: str, fn: SubstitutionHandler) -> None:
    """Register a named placeholder resolver.

    ``name`` is the provider namespace (for example ``"aws"``). ``fn``
    receives one placeholder string and returns the resolved credential
    material. Handlers must run at credential loading time, before any
    request signer computes a body/header/query signature.
    """

    if not name:
        raise ValueError("substitution handler name must be non-empty")
    if not callable(fn):
        raise TypeError("substitution handler must be callable")
    _handlers[name] = fn


def clear_substitution_handlers() -> None:
    """Clear registered handlers. Public for test isolation."""

    _handlers.clear()


def substitute(name: str, placeholder: str) -> str:
    try:
        handler = _handlers[name]
    except KeyError as exc:
        raise SubstitutionHandlerError(
            f"no substitution handler registered for {name!r}"
        ) from exc
    return handler(placeholder)


def aws_credentials_from_placeholders(
    *,
    access_key_id: str,
    secret_access_key: str,
    session_token: str | None = None,
) -> AwsCredentials:
    """Resolve AWS credentials before SigV4 signing.

    Cloud SDK integrations adapt this result into their native
    credential-provider shape:

    - Python: ``botocore.credentials.Credentials``
    - TypeScript: ``@aws-sdk/credential-providers`` provider result
    - Rust: ``aws_config::SdkConfig`` credential provider
    """

    resolved_token = None
    if session_token is not None:
        resolved_token = substitute("aws", session_token)
    return AwsCredentials(
        access_key_id=substitute("aws", access_key_id),
        secret_access_key=substitute("aws", secret_access_key),
        session_token=resolved_token,
    )


def is_placeholder(value: str) -> bool:
    return value.startswith("mvm-secret://")


def constant_time_eq(left: str, right: str) -> bool:
    return compare_digest(left.encode("utf-8"), right.encode("utf-8"))
